From 665c72e17591499bf9f5c054b19d4a53087707d2 Mon Sep 17 00:00:00 2001
From: Jonas Gorski <jonas.gorski@bisdn.de>
Date: Wed, 13 Oct 2021 12:30:26 +0200
Subject: [PATCH 08/21] gpio: allow xgs-iproc for ARCH_XGS_IPROC as well

Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 drivers/gpio/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpio/Kconfig b/drivers/gpio/Kconfig
index d93cd4f722b4..6e4e5a4e55a7 100644
--- a/drivers/gpio/Kconfig
+++ b/drivers/gpio/Kconfig
@@ -209,10 +209,10 @@ config GPIO_BCM_KONA
 
 config GPIO_BCM_XGS_IPROC
 	tristate "BRCM XGS iProc GPIO support"
-	depends on OF_GPIO && (ARCH_BCM_IPROC || COMPILE_TEST)
+	depends on OF_GPIO && (ARCH_BCM_IPROC || ARCH_XGS_IPROC || COMPILE_TEST)
 	select GPIO_GENERIC
 	select GPIOLIB_IRQCHIP
-	default ARCH_BCM_IPROC
+	default ARCH_BCM_IPROC || ARCH_XGS_IPROC
 	help
 	  Say yes here to enable GPIO support for Broadcom XGS iProc SoCs.
 
-- 
2.47.1

