From e86c243a3c131864ee1876045ccf2ddbc79d2376 Mon Sep 17 00:00:00 2001
From: Jonas Gorski <jonas.gorski@bisdn.de>
Date: Thu, 21 Jul 2022 11:03:32 +0200
Subject: [PATCH 12/20] mdio: allow mdio-bcm-iproc for ARCH_XGS_IPROC as well

Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 drivers/net/mdio/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/mdio/Kconfig b/drivers/net/mdio/Kconfig
index bfa16826a6e1..2df681cbc507 100644
--- a/drivers/net/mdio/Kconfig
+++ b/drivers/net/mdio/Kconfig
@@ -85,9 +85,9 @@ config MDIO_BITBANG
 
 config MDIO_BCM_IPROC
 	tristate "Broadcom iProc MDIO bus controller"
-	depends on ARCH_BCM_IPROC || COMPILE_TEST
+	depends on ARCH_BCM_IPROC || ARCH_XGS_IPROC || COMPILE_TEST
 	depends on HAS_IOMEM && OF_MDIO
-	default ARCH_BCM_IPROC
+	default ARCH_BCM_IPROC || ARCH_XGS_IPROC
 	help
 	  This module provides a driver for the MDIO busses found in the
 	  Broadcom iProc SoC's.
-- 
2.38.1

