diff --git a/arch/arm/mm/init.c b/arch/arm/mm/init.c
index a42e4cd11db2..65abb2eec238 100644
--- a/arch/arm/mm/init.c
+++ b/arch/arm/mm/init.c
@@ -430,6 +430,16 @@ static void update_sections_early(struct section_perm perms[], int n)
 	for_each_process(t) {
 		if (t->flags & PF_KTHREAD)
 			continue;
+
+		/*
+		 * A process in getting shut down state (PF_EXITING), with its
+		 * task mmu pointer (mm) being NULL, causes hang in
+		 * set_section_perms().
+		 * NOTE: update_sections_early() runs if CONFIG_DEBUG_RODATA=y
+		 */
+		if (t->flags & PF_EXITING)
+			continue;
+
 		for_each_thread(t, s)
 			if (s->mm)
 				set_section_perms(perms, n, true, s->mm);
-- 
2.43.0

