diff --git a/drivers/net/ethernet/intel/igb/igb_main.c b/drivers/net/ethernet/intel/igb/igb_main.c
index ed7e667d7eb2..6ae4712b1a7c 100644
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -3407,6 +3407,8 @@ static int igb_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 	if (ret_val)
 		strcpy(part_str, "Unknown");
 	dev_info(&pdev->dev, "%s: PBA No: %s\n", netdev->name, part_str);
+	dev_info(&pdev->dev, "%s: mac_type: %i phy_type: %i\n", netdev->name,
+		hw->mac.type, hw->phy.type);
 	dev_info(&pdev->dev,
 		"Using %s interrupts. %d rx queue(s), %d tx queue(s)\n",
 		(adapter->flags & IGB_FLAG_HAS_MSIX) ? "MSI-X" :
